hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/MoveManagerImpl.java

```diff
@@ -0,0 +1,462 @@
+      if (opType.equals(PendingOpType.ADD) && mdnp.getTgt().equals(dn)) {
+        if (pendingMoveFuture.containsKey(containerID)) {
+          try {
+            handleSuccessfulAdd(containerID);
+          } catch (ContainerNotFoundException | NodeNotFoundException |
+                   ContainerReplicaNotFoundException e) {
+            LOG.warn("Can not handle successful Add for move", e);
+          }
+        } else {
+          LOG.warn("No matching entry found in pendingMoveFuture for " +
+              "containerID {}. Should not happen", containerID);
+          completeMove(containerID, null);
+        }
+      } else if (
+          opType.equals(PendingOpType.DELETE) && mdnp.getSrc().equals(dn)) {
+        completeMove(containerID, MoveResult.COMPLETED);
+      }
+    }
+  }
+
+   /**
+   * Notify Move Manager that a container op has been Expired.
+   *
+   * @param containerReplicaOp ContainerReplicaOp
+   * @param containerID ContainerID for which to complete
+   */
+  private void notifyContainerOpExpired(ContainerReplicaOp containerReplicaOp,
+      ContainerID containerID) {
+    if (!running) {
+      return;
+    }
+
+    MoveDataNodePair mdnp = pendingMoveOps.get(containerID);
+    if (mdnp != null) {
+      PendingOpType opType = containerReplicaOp.getOpType();
+      DatanodeDetails dn = containerReplicaOp.getTarget();
+      if (opType.equals(PendingOpType.ADD) && mdnp.getTgt().equals(dn)) {
+        completeMove(containerID, MoveResult.REPLICATION_FAIL_TIME_OUT);
+      } else if (opType.equals(PendingOpType.DELETE) &&
+          mdnp.getSrc().equals(dn)) {
+        completeMove(containerID, MoveResult.DELETION_FAIL_TIME_OUT);
+      }
+    }
+  }
+
+  private void handleSuccessfulAdd(final ContainerID cid)
+      throws ContainerNotFoundException,
+      ContainerReplicaNotFoundException, NodeNotFoundException {
+    MoveDataNodePair movePair = pendingMoveOps.get(cid);
+    if (movePair == null) {
+      return;
+    }
+    final DatanodeDetails src = movePair.getSrc();
+    Set<ContainerReplica> currentReplicas = containerManager
+        .getContainerReplicas(cid);
+    if (currentReplicas.stream()
+        .noneMatch(r -> r.getDatanodeDetails().equals(src))) {
+      // if the target is present but source disappears somehow,
+      // we can consider move is successful.
+      completeMove(cid, MoveResult.COMPLETED);
+      return;
+    }
+
+    final NodeStatus nodeStatus = replicationManager.getNodeStatus(src);
+    if (nodeStatus.getOperationalState()
+        != HddsProtos.NodeOperationalState.IN_SERVICE) {
+      completeMove(cid, MoveResult.DELETION_FAIL_NODE_NOT_IN_SERVICE);
+      return;
+    }
+    if (!nodeStatus.isHealthy()) {
+      completeMove(cid, MoveResult.DELETION_FAIL_NODE_UNHEALTHY);
+      return;
+    }
+
+    Set<ContainerReplica> futureReplicas = currentReplicas.stream()
+        .filter(r -> !r.getDatanodeDetails().equals(src))
+        .collect(Collectors.toSet());
```

Comment:
Nit: we can avoid both streams by using `Collection#removeIf`:

```java
    Set<ContainerReplica> futureReplicas = new HashSet<>(currentReplicas);
    boolean found = futureReplicas.removeIf(r -> r.getDatanodeDetails().equals(src));
    if (!found) {
      // if the target is present but source disappears somehow,
      // we can consider move is successful.
      completeMove(cid, MoveResult.COMPLETED);
      return;
    }
```

Reply:
Good idea.